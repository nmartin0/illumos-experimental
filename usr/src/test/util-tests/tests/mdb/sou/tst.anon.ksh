#!/usr/bin/ksh
#
# This file and its contents are supplied under the terms of the
# Common Development and Distribution License ("CDDL"), version 1.0.
# You may only use this file in accordance with the terms of version
# 1.0 of the CDDL.
#
# A full copy of the text of the CDDL should have accompanied this
# source.  A copy of the CDDL is also available via the Internet at
# http://www.illumos.org/license/CDDL.
#

#
# Copyright 2025 Oxide Computer Company
#

#
# Test our ability to recurse through and print information about anonymous
# structs and unions. Effectively, this is about the resolution capabilities of
# various dcmds. This is paired with the "anon" program. Note, we have to be
# very careful to avoid printing addresses into the output data we're comparing
# against. Make sure to use ::printf for strings rather than ::print or './s'
#

set -o pipefail

tst_root="$(dirname $0)/.."
tst_prog="$tst_root/progs/anon"

cat </dev/null > $ODIR/stdout

#
# Begin with basic offsetof logic on types. This ensures that embedded anonymous
# entities are at the correct offset.
#
$MDB -e "::offsetof struct foo foo" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct bar bar" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct bar bar_foo" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct bar bar_int" $tst_prog >> $ODIR/stdout

$MDB -e "::offsetof struct baz baz_str" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct baz baz_anon" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct baz baz_int" $tst_prog >> $ODIR/stdout

$MDB -e "::offsetof struct foobar foobar_int" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar foo" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar bar" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar baz" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar foobar_anon" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar a" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar b" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar c" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar d" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar e" $tst_prog >> $ODIR/stdout
$MDB -e "::offsetof struct foobar f" $tst_prog >> $ODIR/stdout

#
# Print ::sizeof to make sure that this aligns with offsetof data.
#
$MDB -e "::sizeof struct foo" $tst_prog >> $ODIR/stdout
$MDB -e "::sizeof struct bar" $tst_prog >> $ODIR/stdout
$MDB -e "::sizeof struct baz" $tst_prog >> $ODIR/stdout
$MDB -e "::sizeof struct foobar" $tst_prog >> $ODIR/stdout

#
# Print the overall structure to make sure we get the <anon> entries we expect.
#
$MDB -e "foo::print" $tst_prog >> $ODIR/stdout
$MDB -e "foo::print foo" $tst_prog >> $ODIR/stdout
$MDB -e "bar::printf \"%s\\n\" struct bar bar" $tst_prog >> $ODIR/stdout
$MDB -e "bar::print bar_foo" $tst_prog >> $ODIR/stdout
$MDB -e "bar::print bar_foo.foo" $tst_prog >> $ODIR/stdout
$MDB -e "bar::print bar_foo.foo bar_int" $tst_prog >> $ODIR/stdout
$MDB -e "baz::printf \"%s\\n\" struct baz baz_str" $tst_prog >> $ODIR/stdout
$MDB -e "baz::printf \"0x%x 0x%x\\n\" struct baz baz_anon baz_int" \
    $tst_prog >> $ODIR/stdout
$MDB -e "foobar::printf \"0x%x 0x%x 0x%x 0x%x 0x%x 0x%x\\n\" struct foobar \
    a b c d e f" $tst_prog >> $ODIR/stdout
$MDB -e "foobar::printf \"%s %s\\n\" struct foobar bar.bar baz.baz_str " \
    $tst_prog >> $ODIR/stdout
$MDB -e "foobar::print struct foobar foobar_int" $tst_prog >> $ODIR/stdout
$MDB -e "foobar::printf \"0x%x 0x%x 0x%x 0x%x\\n\" struct foobar bar.bar_foo.foo \
    bar.bar_int baz.baz_anon foobar_anon" $tst_prog >> $ODIR/stdout

#
# The stringless structure has no strings, so we can print it without worrying
# about addresses and make sure we get the appropriate anon names.
#
$MDB -e "stringless::print" $tst_prog >> $ODIR/stdout
$MDB -e "stringless::print -t" $tst_prog >> $ODIR/stdout

#
# Use the direct offset specification syntax to print a few things as
# well. This only works with ::print.
#
$MDB -e "bar::print \$[8]" $tst_prog >> $ODIR/stdout
$MDB -e "foobar::print \$[20] \$[2c]" $tst_prog >> $ODIR/stdout
