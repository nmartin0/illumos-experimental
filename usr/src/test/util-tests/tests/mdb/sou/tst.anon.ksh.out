offsetof (struct foo, foo) = 0, sizeof (...->foo) = 4
offsetof (struct bar, bar) = 0, sizeof (...->bar) = 8
offsetof (struct bar, bar_foo) = 8, sizeof (...->bar_foo) = 4
offsetof (struct bar, bar_int) = 8, sizeof (...->bar_int) = 4
offsetof (struct baz, baz_str) = 0, sizeof (...->baz_str) = 8
offsetof (struct baz, baz_anon) = 8, sizeof (...->baz_anon) = 4
offsetof (struct baz, baz_int) = 0x10, sizeof (...->baz_int) = 4
offsetof (struct foobar, foobar_int) = 0, sizeof (...->foobar_int) = 4
offsetof (struct foobar, foo) = 8, sizeof (...->foo) = 4
offsetof (struct foobar, bar) = 8, sizeof (...->bar) = 0x10
offsetof (struct foobar, baz) = 8, sizeof (...->baz) = 0x18
offsetof (struct foobar, foobar_anon) = 0x10, sizeof (...->foobar_anon) = 4
offsetof (struct foobar, a) = 0x20, sizeof (...->a) = 4
offsetof (struct foobar, b) = 0x24, sizeof (...->b) = 4
offsetof (struct foobar, c) = 0x28, sizeof (...->c) = 4
offsetof (struct foobar, d) = 0x2c, sizeof (...->d) = 4
offsetof (struct foobar, e) = 0x2c, sizeof (...->e) = 4
offsetof (struct foobar, f) = 0x2c, sizeof (...->f) = 4
sizeof (struct foo) = 4
sizeof (struct bar) = 0x10
sizeof (struct baz) = 0x18
sizeof (struct foobar) = 0x30
{
    foo = 0x2a
}
foo = 0x2a
hello world
bar_foo = {
    bar_foo.foo = 0x7777
}
bar_foo.foo = 0x7777
bar_foo.foo = 0x7777
bar_int = 0x7777
It's a trap?!
0x9999 0xfffffffc
0x9876 0x12345 0xb22b 0xfdcba 0xfdcba 0xfdcba
Elbereth Elbereth
foobar_int = 0xb295
0x7777 0x7777 0x7777 0x7777
{
    <anon> = {
        <anon> = {
            life = 0xaa7777aa
            hope = 0x339999ee
        }
        dreams = 0xaa7777aa
        <anon> = {
            k = '\252'
            e = 0x77aa
            f = 0xaa7777aa
            a = 0x339999eeaa7777aa
        }
        destroy = +3.9829317e-60
    }
}
struct stringless {
    union  <anon> = {
        struct  <anon> = {
            int life = 0xaa7777aa
            int hope = 0x339999ee
        }
        int dreams = 0xaa7777aa
        union  <anon> = {
            char k = '\252'
            short e = 0x77aa
            int f = 0xaa7777aa
            long a = 0x339999eeaa7777aa
        }
        double destroy = +3.9829317e-60
    }
}
<anon> = {
    <anon>.bar_foo = {
        foo = 0x7777
    }
    <anon>.bar_int = 0x7777
}
<anon> = {
    <anon>.a = 0x9876
    <anon>.b = 0x12345
    <anon>.c = 0xb22b
}
<anon> = {
    <anon>.d = 0xfdcba
    <anon>.e = 0xfdcba
    <anon>.f = 0xfdcba
}
